---
title: "Assignment 1"
author: "Yasamin Nouri Jelyani"
output:
  html_document:
    toc: true
    toc_depth: 2
    df_print: paged
bibliography: a2.bib
csl: biomed-central.csl
nocite: '@*'
---

#Introduction

A summary of the paper and datasets

# Download the Data

First, I downloaded the GEOmetaDB, and got access to the meta data:
```{r eval=FALSE}

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

if (!requireNamespace("GEOmetadb", quietly = TRUE))
    BiocManager::install("GEOmetadb")

if(!requireNamespace("knitr", quietly=TRUE))
  install.packages("knitr")

if(! requireNamespace("edgeR", quietly=TRUE))
  BiocManager::install("edgeR")

if (!requireNamespace("GEOquery", quietly=TRUE))
  BiocManager::install("GEOquery")

library(GEOmetadb)

if( !file.exists("GEOmetadb.sqlite") ) {
    demo_sqlfile <- getSQLiteFile(destdir = getwd(), destfile = "GEOmetadb.sqlite.gz", type = "demo")
} else {
    demo_sqlfile <- "GEOmetadb.sqlite"
    
}

file.info(demo_sqlfile)

con <- dbConnect(SQLite(), demo_sqlfile)

```

Query of the database:
```{r eval=FALSE}
sql <- paste("SELECT DISTINCT gse.title,gse.gse, gpl.title,",
             " gse.submission_date,",
             " gse.supplementary_file", #<<
             "FROM",
             "  gse JOIN gse_gpl ON gse_gpl.gse=gse.gse",
             "  JOIN gpl ON gse_gpl.gpl=gpl.gpl",
             "WHERE",
             "  gse.submission_date > '2017-01-01' AND",
             "  gse.title LIKE '%cancer%' AND", 
             "  gpl.organism LIKE '%Homo sapiens%' AND",
             "  gpl.technology LIKE '%high-throughput sequencing%' ",
             "  ORDER BY gse.submission_date DESC",sep=" ")

rs <- dbGetQuery(con,sql)
counts_files <- rs$supplementary_file[grep(rs$supplementary_file,
                              pattern = "count|cnt",ignore.case = TRUE)]

series_of_interest <- rs$gse[grep(rs$supplementary_file,
                              pattern = "count|cnt",ignore.case = TRUE)]

shortened_filenames <- unlist(lapply(counts_files,
              FUN = function(x){x <- unlist(strsplit(x,";")) ;
              x <- x[grep(x,pattern= "count|cnt",ignore.case = TRUE)];
                tail(unlist(strsplit(x,"/")),n=1)}))
shortened_filenames[1:10]
rs <- dbGetQuery(con,sql)
knitr::kable(rs, format = "html")

```

Open the data:
```{r message=FALSE}
sfiles = GEOquery::getGEOSuppFiles('GSE117452')

#if the file exists, open it
fnames <- "GSE117452/GSE117452_expn_matrix_mimat.txt.gz"
if (! file.exists(fnames)){
  gse <- GEOquery::getGEO("GSE117452", GSEMatrix=FALSE)
  sfiles = GEOquery::getGEOSuppFiles("GSE117452")
  fnames = rownames(sfiles)
}
```



Geo Description of dataset:
```{r message=FALSE}
gse <- GEOquery::getGEO("GSE117452",GSEMatrix=FALSE)

knitr::kable(data.frame(head(GEOquery::Meta(gse))), format = "html")
```

Platform title:
```{r}
current_gpl <- names(GEOquery::GPLList(gse))[1]
current_gpl_info <- GEOquery::Meta(GEOquery::getGEO(current_gpl))
current_gpl_info$title
```


```{r}
current_gpl_info$last_update_date
```


```{r}
current_gpl_info$organism
```
Getting the expression data
```{r message=FALSE}
sfiles = GEOquery::getGEOSuppFiles('GSE117452')

fnames = rownames(sfiles)

#69 columns, there are 68 samples (58 breast cancer patients 
#(29 AA and 29 EA women) and 10 (5 AA and 5 EA) healthy women.)

#589 rows (Genes) differentially expressed microRNA

ca125_exp = read.delim(fnames[1],header=TRUE, check.names = FALSE)


knitr::kable(ca125_exp[1:15,1:5], format = "html")
```

cleaning the data
```{r}
dim(ca125_exp)

colnames(ca125_exp)
```

#Assess the data:
The data does not have duplicates.
```{r}
num_samples <- dim(ca125_exp)[2] - 1 # The number of rows correspond to the number of columns minus 1 for the gene name colummn 68
num_genes <- dim(ca125_exp)[1] # The number of genes analysed 2576 genes
num_unique_genes <- length(unique(ca125_exp$Gene)) # How many unique genes are measured: 2576
num_lack_identifier <- sum(is.na(ca125_exp$Gene)) # How many rows do not have an gene identifier: 0

```

There are no repetative genes.

```{r}
summarized_gene_counts <- sort(table(ca125_exp$Gene),
                               decreasing = TRUE)
```

```{r}
knitr::kable(table(ca125_exp$Gene)[1:3], format="html")
```

```{r}
#translate out counts into counts per million using 
#the edgeR package function cpm
cpms = edgeR::cpm(ca125_exp[,2:ncol(ca125_exp)])
rownames(cpms) <- ca125_exp[,1]
# get rid of low counts
keep = rowSums(cpms >1) >=10
ca125_exp_filtered = ca125_exp[keep,]
```

```{r}
dim(ca125_exp_filtered)
```
# Mapping

```{r warning=FALSE}
# do not have space for all 
data2plot <- log2(edgeR::cpm(ca125_exp_filtered[,2:20]))
boxplot(data2plot, xlab = "Samples", ylab = "log2 CPM", 
        las = 2, cex = 0.5, cex.lab = 0.5,
        cex.axis = 0.5, main = "CA125 RNASeq Samples")
#draw the median on each box plot
abline(h = median(apply(data2plot, 2, median)), 
       col = "green", lwd = 0.6, lty = "dashed")
```



```{r}
counts_density <- apply(log2(edgeR::cpm(ca125_exp_filtered[,2:ncol(ca125_exp)])), 
                        2, density)
  #calculate the limits across all the samples
    xlim <- 0; ylim <- 0
    for (i in 1:length(counts_density)) {
      xlim <- range(c(xlim, counts_density[[i]]$x)); 
      ylim <- range(c(ylim, counts_density[[i]]$y))
    }
    cols <- rainbow(length(counts_density))
    ltys <- rep(1, length(counts_density))
    #plot the first density plot to initialize the plot
    plot(counts_density[[1]], xlim=xlim, ylim=ylim, type="n", 
         ylab="Smoothing density of log2-CPM", 
         main="", cex.lab = 0.85)
    #plot each line
    for (i in 1:length(counts_density)) 
      lines(counts_density[[i]], col=cols[i], lty=ltys[i])
    #create legend
    legend("topright", colnames(data2plot),  
           col=cols, lty=ltys, cex=0.75, 
           border ="blue",  text.col = "green4", 
           merge = TRUE, bg = "gray90")
```
```{r}
limma::plotMA(log2(ca125_exp[,c(3,4)]), ylab="M - ratio log expression", 
       main="CA125 + vs - - example")
```


Disconnect from the database:
```{r eval=FALSE}

dbDisconnect(con)

```

# References

